salt/pillar/rethinkdb_pillar.py

```diff
@@ -0,0 +1,160 @@
+
+__opts__ = {
+    'rethinkdb.host': 'salt',
+    'rethinkdb.port': '28015',
+    'rethinkdb.database': 'salt',
+    'rethinkdb.username': None,
+    'rethinkdb.password': None
+}
+
+
+def __virtual__():
+    if not HAS_RETHINKDB:
+        return False
+    return True
+
+
+# Configure logging
+log = logging.getLogger(__name__)
+
+
+def ext_pillar(minion_id,
+               pillar,
+               table='pillar',
+               id_field=None,
+               field=None,
+               pillar_key=None):
+    '''
+    Collect minion external pillars from a RethinkDB database
+
+Arguments:
+    * `table`: The RethinkDB table containing external pillar information.
+      Defaults to ``'pillar'``
+    * `id_field`: Field in document containing the minion id.
+      If blank then we assume the table index matches minion ids
+    * `field`: Specific field in the document used for pillar data, if blank
+      then the entire document will be used
+    * `pillar_key`: The salt-master will nest found external pillars under
+      this key before merging into the minion pillars. If blank, external
+      pillars will be merged at top level
+    '''
+    host = __opts__['rethinkdb.host']
+    port = __opts__['rethinkdb.port']
+    database = __opts__['rethinkdb.database']
+    username = __opts__['rethinkdb.username']
+    password = __opts__['rethinkdb.password']
+
+    log.debug('Connecting to {0}:{1} as user \'{2}\' for RethinkDB ext_pillar'
+              .format(host, port, username))
+
+    # Connect to the database
+    conn = r.connect(host=host,
+                     port=port,
+                     db=database,
+                     user=username,
+                     password=password)
+
+    data = None
+
+    try:
+
+        if id_field:
+            log.debug('ext_pillar.rethinkdb: looking up pillar. '
+                      'table: {0}, field: {1}, minion: {2}'.format(
+                          table, id_field, minion_id))
+
+            if field:
+                data = r.table(table).filter(
+                    {id_field: minion_id}).pluck(field).run(conn)
+            else:
+                data = r.table(table).filter({id_field: minion_id}).run(conn)
+
+        else:
+            log.debug('ext_pillar.rethinkdb: looking up pillar. '
+                      'table: {0}, field: id, minion: {1}'.format(
+                          table, minion_id))
+
+            if field:
+                data = r.table(table).get(minion_id).pluck(field).run(conn)
+            else:
+                data = r.table(table).get(minion_id).run(conn)
+
+    finally:
+        if conn.is_open():
+            conn.close()
+
+    if data.items:
+
+        # Return nothing if multiple documents are found for a minion
+        if len(data.items) > 1:
+            log.error('ext_pillar.rethinkdb: ambiguous documents found for '
+                      'minion {0}'.format(minion_id))
+            return {}
+
+        else:
+            for document in data:
```

Comment:
I'm not sure I totally follow the reasoning for this loop. The way I read this, if it's > 1 then the above block will execute but if it's 1 or 0, then we'll go here but it's not clear why we're looping for 1 | 0 items. Am I missing something here?

Reply:
You're quite correct. I was going by the RethinkDB examples I read some time ago that dealt with iterating over results in a cursor. I updated the pull request and replaced the for loop with pop()